src/agent/src/rpc.rs

```diff
@@ -291,35 +285,36 @@ impl AgentService {
         &self,
         req: protocols::agent::RemoveContainerRequest,
     ) -> Result<()> {
-        let cid = req.container_id.clone();
+        let cid = req.container_id;
 
         if req.timeout == 0 {
             let mut sandbox = self.sandbox.lock().await;
-
             sandbox.bind_watcher.remove_container(&cid).await;
-
             sandbox
                 .get_container(&cid)
                 .ok_or_else(|| anyhow!("Invalid container id"))?
                 .destroy()
                 .await?;
-
             remove_container_resources(&mut sandbox, &cid)?;
-
             return Ok(());
         }
 
         // timeout != 0
+        let (tx, rx) = tokio::sync::oneshot::channel::<i32>();
         let s = self.sandbox.clone();
         let cid2 = cid.clone();
-        let (tx, rx) = tokio::sync::oneshot::channel::<i32>();
 
         let handle = tokio::spawn(async move {
             let mut sandbox = s.lock().await;
             if let Some(ctr) = sandbox.get_container(&cid2) {
-                ctr.destroy().await.unwrap();
+                if let Err(e) = ctr.destroy().await {
```

Comment:
Good point, I have solved 1 & 3, and 2 is unsolvable.
I have tried to solve 2 before but found that will cause lifetime issue, then give up.

Reply:
```diff
         // timeout != 0
         let s = self.sandbox.clone();
         let cid2 = cid.clone();
-        let (tx, rx) = tokio::sync::oneshot::channel::<i32>();
+        let (tx, rx) = tokio::sync::oneshot::channel::<Result<()>>();
 
         let handle = tokio::spawn(async move {
             let mut sandbox = s.lock().await;
-            if let Some(ctr) = sandbox.get_container(&cid2) {
-                ctr.destroy().await.unwrap();
-                sandbox.bind_watcher.remove_container(&cid2).await;
-                tx.send(1).unwrap();
+
+            sandbox.bind_watcher.remove_container(&cid2).await;
+
+            let ret = match sandbox.get_container(&cid2) {
+                None => Err(anyhow!("Invalid container id")),
+                Some(ctr) => ctr.destroy().await,
             };
+
+            if let Err(e) = tx.send(ret) {
+                warn!(sl(), "failed to send stopping container {}, {:?}", cid2, e);
+            }
         });
```

Reply:
Thanks for the suggestion, I have simplified the code further. Please help to review the version at: https://github.com/kata-containers/kata-containers/pull/7552/commits/8e70fe01f74e9f0f4eb4deea91b79cec7ee03974

Reply:
Yeah, I thought about that too but that changes the semantics of the function.

Currently, if we hit a timeout, we detach from the running task and return a failure -- this means that we simply give up on waiting for the removal to complete, but the removal itself is allowed to continue even if/when we fail the request.

With your changes in https://github.com/kata-containers/kata-containers/commit/8e70fe01f74e9f0f4eb4deea91b79cec7ee03974, once we hit the timeout, we *cancel* the task. That is, the removal is not allowed to proceed anymore.

IOW, if there was a legitimate reason for the task to take longer than the timeout (e.g., the lock was highly-contended, but would eventually be acquired), then the removal would still succeed. In the latest code, it isn't the case anymore, we'll simply abort the task and the removal won't be able to complete.

Reply:
We are going to about the `JoinHandle` instead of the `task`, so the task should run to complete even we drop the `JoinHandle`.

Reply:
You're right. In that case, I would remove all `match` statements in favor of `?`:

```diff
@@ -312,27 +312,20 @@ impl AgentService {
         // timeout != 0
         let s = self.sandbox.clone();
         let cid2 = cid.clone();
-        let (tx, rx) = tokio::sync::oneshot::channel::<i32>();
 
         let handle = tokio::spawn(async move {
             let mut sandbox = s.lock().await;
-            if let Some(ctr) = sandbox.get_container(&cid2) {
-                ctr.destroy().await.unwrap();
-                sandbox.bind_watcher.remove_container(&cid2).await;
-                tx.send(1).unwrap();
-            };
+            sandbox.bind_watcher.remove_container(&cid2).await;
+            sandbox
+                .get_container(&cid2)
+                .ok_or_else(|| anyhow!("Invalid container id"))?
+                .destroy()
+                .await
         });

-        if tokio::time::timeout(Duration::from_secs(req.timeout.into()), rx)
+        tokio::time::timeout(Duration::from_secs(req.timeout.into()), handle)
             .await
-            .is_err()
-        {
-            return Err(anyhow!(nix::Error::ETIME));
-        }
-
-        if handle.await.is_err() {
-            return Err(anyhow!(nix::Error::UnknownErrno));
-        }
+            .map_err(|_| anyhow!(nix::Error::ETIME))???;
```

Reply:
Ok. It takes too much time to run CI tasks. So how about merging this PR first and I will submit a minor patch to follow on?

Reply:
Nevermind, I will fold this into my PR once I rebase on top of yours.

Let's get this merged so that I can rebase mine and shake out the common parts.